dbms/src/Columns/Collator.cpp

```diff
@@ -26,16 +30,75 @@ namespace DB
     }
 }
 
+
+std::unique_ptr<AvailableCollationLocales> AvailableCollationLocales::instance_impl;
+std::once_flag AvailableCollationLocales::init_flag;
+
+void AvailableCollationLocales::init()
+{
+    instance_impl = std::make_unique<AvailableCollationLocales>();
+#if USE_ICU
+    size_t available_locales_count = ucol_countAvailable();
+    for (size_t i = 0; i < available_locales_count; ++i)
+    {
+        std::string locale_name = ucol_getAvailable(i);
+        UChar lang_buffer[128]; /// 128 is enough for language name
+        char normal_buf[128];
+        UErrorCode status = U_ZERO_ERROR;
+        /// All names will be in English language
+        size_t lang_length = uloc_getDisplayLanguage(locale_name.c_str(), "en", lang_buffer, 128, &status);
+        if (U_FAILURE(status))
+            instance_impl->available_collation_locales.push_back(LocaleAndLanguage{locale_name, "unknown"});
+        else
+        {
+            u_UCharsToChars(lang_buffer, normal_buf, lang_length);
+            LocaleAndLanguage result{locale_name, std::string(normal_buf, lang_length)};
+            instance_impl->available_collation_locales.push_back(result);
+        }
+    }
+
+    auto comparator = [] (const LocaleAndLanguage & f, const LocaleAndLanguage & s) { return f.locale_name < s.locale_name; };
+    std::sort(instance_impl->available_collation_locales.begin(), instance_impl->available_collation_locales.end(), comparator);
+#endif
+}
+
+AvailableCollationLocales & AvailableCollationLocales::instance()
+{
+    std::call_once(init_flag, AvailableCollationLocales::init);
+    return *instance_impl;
+}
+
+const std::vector<AvailableCollationLocales::LocaleAndLanguage> & AvailableCollationLocales::getAvailableCollations() const
+{
+    return available_collation_locales;
+}
+
+bool AvailableCollationLocales::isCollationSupported(const std::string & s) const
+{
+    std::string lower = Poco::toLower(s);
+    for (const auto & locale_and_lang : available_collation_locales)
```

Comment:
Not important, but probably we can use hash table to answer faster.

Reply:
I'm not sure about it (size of the vector is extremely small). And also we will have to save this data is a vector as well, because we need original names in `system.collations`.

Reply:
Could you please add a comment? Otherwise people who read code will stop here and ask questions.